### DISCUSSION TITLE
Multiple releases on the same line from a monorepo

### DISCUSSION THREAD

**Discussion by ChipWolf** [renovatebot/renovate#14759] [Request Help] [answered]
Hey! 👋 I'm looking for a little help with this scenario, please:

We maintain a monorepo where we release several packages, the GitHub releases are tagged like this:
```
pkg-foo-v1.0.1
bar-v3.2.5
the-other-qux-v12.4.0
etc.
```

---

In another repo _(where Renovate runs)_, we depend on those packages, but they're defined a little unconventionally:
_(the +xx signifies a locality)_

```yaml
env:
  UPSTREAM_PACKAGES: "the-other-qux-v12.4.0,pkg-foo-v1.0.1,bar-v3.2.5+de"
```
or
```yaml
env:
  UPSTREAM_PACKAGES: "bar-v3.2.1+en,pkg-foo-v1.0.1"
```
or
```yaml
env:
  UPSTREAM_PACKAGES: "pkg-foo-v1.0.1"
```

---

Previously, our upstream repo only contained one package (pkg-foo), the tags were not prefixed, and each of our deployments required only one package _(like the third example above)_; so this Renovate config worked well:
```json
"regexManagers": [
  {
    "fileMatch": ["./.*\\.yaml$"],
    "matchStrings": [
      "UPSTREAM_PACKAGES: \"?pkg-foo-(?<currentValue>.+)\\+[a-z]{2}\"?\n",
    ],
    "depNameTemplate": "our-org/monorepo",
    "datasourceTemplate": "github-releases"
  }
]
```

However, now we must maintain our packages in a monorepo, and we must reference multiple different packages on a single line in our downstream repo; my Renovate expertise breaks down here. Any help would be greatly appreciated.

Thanks in advance!

**Comment by rarkins** [comment]
Just a wild guess, I'm wondering if something like this works:
- extract both depName and packageName using the regex manager. DepName will be the pkg-foo part
- use extractVersion for the datasource with {{depName}}- inside it

**Reply by ChipWolf** [reply]
I've managed to get this far:

```regex
(?=(?:"|,)(?<depName>.+?)-(?<currentValue>v[0-9].+?)(?:\+[a-z]{2})?(?:"|,))
```
![](https://i.cwlf.uk/ORXok)

```json
"packageRules": [
  {
    "datasources": ["github-releases"],
    "extractVersion": "^(?<depName>.+?)-(?<version>v[0-9].*)",
  }
],
"regexManagers": [
  {
    "fileMatch": ["./.*\\.yaml$"],
    "matchStrings": [
      "(?=(?:\"|,)(?<depName>.+?)-(?<currentValue>v[0-9].+?)(?:\\+[a-z]{2})?(?:\"|,))",
    ],
    "depNameTemplate": "our-org/monorepo",
    "datasourceTemplate": "github-releases"
  },
],
```

However, Renovate doesn't seem to be able to validate the regex. Any tips here? Many thanks
```json
  INFO: Repository has invalid config (repository=our-org/downstream)
       "error": {
         "validationSource": "(?=(?:\"|,)(?<depName>.+?)-(?<currentValue>v[0-9].+?)(?:\\+[a-z]{2})?(?:\"|,))",
         "validationError": "Invalid regular expression: (?=(?:\"|,)(?<depName>.+?)-(?<currentValue>v[0-9].+?)(?:\\+[a-z]{2})?(?:\"|,))",
         "message": "config-validation",
...
```

**Reply by viceice** [reply]
You can't use look ahead / behind, it's blocked in our RE2 regex engine for good reasons

**Reply by HonkingGoose** [reply]
Could we create a log message that explains that look ahead and look behind are not supported? That way our users know what they're doing wrong. 😉

**Reply by ChipWolf** [reply]
I figured lookaheads were not permitted here due to RE2, so I made some further progress:

```json
"packageRules": [
  {
    "datasources": ["github-releases"],
    "groupName": "monorepo",
  },
],
"regexManagers": [
  {
    "depNameTemplate": "our-org/monorepo",
    "datasourceTemplate": "github-releases"
    "extractVersionTemplate": "^(?<depName>.+?)-(?<version>v[0-9].*)",
    "fileMatch": ["./.*\\.yaml$"],
    "matchStrings": [
      "(?:\"|,)(?<depName>.+?)-(?<currentValue>v[0-9].+?)(?:\\.zip|\\+|\")",
    ],
  },
]
```

---

The above works erratically. One pull request is opened to update both dependencies as expected:

| Package | Update | Change |
|---|---|---|
| our-org/monorepo | minor | `v3.2.1` -> `v3.3.1` |
| our-org/monorepo | patch | `v1.1.1` -> `v1.1.2` |

```diff
- UPSTREAM_PACKAGES: "bar-v3.2.1+en.zip,pkg-foo-v1.1.1.zip"
+ UPSTREAM_PACKAGES: "bar-v3.3.1+en.zip,pkg-foo-v1.1.2.zip"
```

---

However, a second pull request is opened with `-major-` in the branch name, with this erroneous change:

| Package | Update | Change |
|---|---|---|
| our-org/monorepo | major | `v1.1.1` -> `v3.3.1` |

```diff
- UPSTREAM_PACKAGES: "bar-v3.2.1+en.zip,pkg-foo-v1.1.1.zip"
+ UPSTREAM_PACKAGES: "bar-v3.2.1+en.zip,pkg-foo-v3.3.1.zip"
```

---

If I add `"separateMajorMinor": "false"` to the packageRule, only one pull request is opened, but with a combination of correctly bumping the version for `bar` but erroneously bumping the version for `pkg-foo` with the same version as `bar`:

| Package | Update | Change |
|---|---|---|
| our-org/monorepo | major | `v1.1.1` -> `v3.3.1` |
| our-org/monorepo | minor | `v3.2.1` -> `v3.3.1` |

```diff
- UPSTREAM_PACKAGES: "bar-v3.2.1+en.zip,pkg-foo-v1.1.1.zip"
+ UPSTREAM_PACKAGES: "bar-v3.3.1+en.zip,pkg-foo-v3.3.1.zip"
```

---

Interestingly, the following 4 diffs are listed in the release notes with the final example:
```
bar-v3.3.0...bar-v3.3.1

bar-v3.2.1...bar-v3.3.0

v1.1.2...bar-v3.3.1

v1.1.1...v1.1.2
```

These diffs indicate it's performing the update correctly for the `pkg-foo` dependency first, then applying the erroneous change after. I also note the tag prefix is missing from these comparisons.

**Comment by ChipWolf** [comment, marked as answer]
This is how I ended up solving this:
```json
"packageRules": [
  {
    "matchDatasources": ["github-releases"],
    "versioning": "regex:^(?<compatibility>.*)-v(?<major>\\d+)(\\.(?<minor>\\d+))(\\.(?<patch>\\d+))$",
  }
],
"regexManagers": [
  {
    "depNameTemplate": "our-org/monorepo",
    "datasourceTemplate": "github-releases"
    "fileMatch": ["./.*\\.yaml$"],
    "matchStrings": [
      "(?:\"|,)(?<currentValue>.+?)(?:\\.zip|\\+|\")",
    ],
  }
]
```

If anyone has a better solution, please let me know. Thanks!

**Reply by ChipWolf** [reply]
An interesting side effect of this, PRs have an extra `v` before the tag name.
`chore(deps): update dependency our-org/monorepo to vfoo-pkg-v3.2.1`

**Reply by agermain** [reply]
Curious, did you mean `matchDatasources` in your `packageRules` block here?

**Reply by rarkins** [reply]
`datasources` is the old name and is automatically migrated to `matchDatasources` internally
